graviti/portex/factory.py

```diff
@@ -43,6 +46,134 @@ def __call__(self, **_: Any) -> Any:
         ...
 
 
+class ImmutableFieldsFactory:
+    """The factory for ImmutableFields.
+
+    Arguments:
+        decl: The decalaration of immutable fields.
+        imports: The :class:`Imports` instance to specify the import scope of the fields.
+
+    """
+
+    def __init__(self, decl: Iterable[Dict[str, Any]], imports: Imports) -> None:
+        self._factories = [FieldFactory(field, imports) for field in decl]
+
+    def __call__(self, **kwargs: Any) -> ImmutableFields:
+        """Apply the input arguments to the ImmutableFields factory.
+
+        Arguments:
+            kwargs: The input arguments.
+
+        Returns:
+            The applied ImmutableFields.
+
+        """
+        return ImmutableFields(
+            filter(
+                bool,
+                (factory(**kwargs) for factory in self._factories),  # type: ignore[misc]
+            )
+        )
+
+
+class BaseImmutableFieldsFactory:
+    """The factory for ConnectedFields.
+
+    Arguments:
+        factory: The factory of immutable fields.
+        internal_kwargs_getter: The method to get the kwargs of the base type.
+
+    """
+
+    def __init__(
+        self,
+        factory: Callable[..., UnionFields],
+        internal_kwargs_getter: Callable[..., Dict[str, Any]],
+    ) -> None:
+        self._factory = factory
+        self._internal_kwargs_getter = internal_kwargs_getter
+
+    def __call__(self, **kwargs: Any) -> ImmutableFields:
+        """Apply the input arguments to the base ImmutableFields factory.
+
+        Arguments:
+            kwargs: The input arguments.
+
+        Returns:
+            The applied ImmutableFields.
+
+        """
+        return self._factory(**self._internal_kwargs_getter(**kwargs))
+
+
+class ConnectedFieldsFactory:
+    """The factory for ConnectedFields.
+
+    Arguments:
+        decl: A dict which indicates a portex type.
+        class_: The internal type.
+        imports: The :class:`Imports` instance to specify the import scope of the template.
+        internal_kwargs_getter: The method to get the kwargs of the internal type.
+
+    """
+
+    def __init__(
+        self,
+        decl: Dict[str, Any],
+        class_: Type[PortexRecordBase],
+        imports: Imports,
+        internal_kwargs_getter: Callable[..., Dict[str, Any]],
+    ) -> None:
+        factories: List[Tuple[Optional[str], Callable[..., UnionFields]]] = []
+        for base_key, base_factory in class_._fields_factory._factories:  # type: ignore[has-type]
+            if base_key is None:  # This means _fields_factory of base class is for ImmutableFields.
```

Comment:
can we use isinstance to check whether the factory is VariableFactory or `ImmutableFieldsFactory`?